File: android/app/src/main/java/com/commissionerscartel/app/feature/coach/CoachViewModel.kt
package com.commissionerscartel.app.feature.coach

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.commissionerscartel.app.data.Config
import com.commissionerscartel.app.data.Session
import io.ktor.client.HttpClient
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.request.header
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

data class CoachTurn(val text: String, val fromCoach: Boolean)

data class CoachState(
    val turns: List<CoachTurn> = emptyList(),
    val busy: Boolean = false,
)

class CoachViewModel : ViewModel() {
    private val http = HttpClient(OkHttp)
    private val json = Json { ignoreUnknownKeys = true }

    private val _state = MutableStateFlow(CoachState())
    val state: StateFlow<CoachState> = _state.asStateFlow()

    private var loadedHistory = false

    /**
     * Brings back what was said before.
     *
     * The conversation lives on the server rather than in this view model, so
     * it survives the app being closed and is the same one the member sees on
     * their other phone.
     */
    fun loadHistory() {
        if (loadedHistory) return
        loadedHistory = true

        viewModelScope.launch {
            val earlier = runCatching { history() }.getOrElse { emptyList() }
            if (earlier.isEmpty()) return@launch
            // Anything typed while this was loading stays where it was, at the
            // end of the conversation.
            _state.value = _state.value.copy(turns = earlier + _state.value.turns)
        }
    }

    private suspend fun history(): List<CoachTurn> {
        val token = Session.accessToken() ?: return emptyList()
        val response = http.get("${Config.supabaseUrl}/rest/v1/coach_messages") {
            header("apikey", Config.supabaseAnonKey)
            header("Authorization", "Bearer $token")
            // Newest first so the limit keeps the recent end of a long
            // conversation; reversed below for reading.
            parameter("select", "role,content,seq")
            parameter("order", "seq.desc")
            parameter("limit", "100")
        }
        return json.parseToJsonElement(response.bodyAsText()).jsonArray
            .map { row ->
                val obj = row.jsonObject
                CoachTurn(
                    text = obj["content"]?.jsonPrimitive?.content.orEmpty(),
                    fromCoach = obj["role"]?.jsonPrimitive?.content == "coach",
                )
            }
            .reversed()
    }

    fun ask(question: String) {
        if (question.isBlank() || _state.value.busy) return

        _state.value = _state.value.copy(
            turns = _state.value.turns + CoachTurn(question, fromCoach = false),
            busy = true,
        )

        viewModelScope.launch {
            val reply = runCatching { request(question) }
                .getOrElse { "Coach Landry isn't answering just now." }
            _state.value = _state.value.copy(
                turns = _state.value.turns + CoachTurn(reply, fromCoach = true),
                busy = false,
            )
        }
    }

    private suspend fun request(question: String): String {
        // The member's own token, not the anon key: the function answers about
        // whoever is asking, so it has to know who that is.
        val token = Session.accessToken() ?: return "Sign in first, under Settings."

        val response = http.post("${Config.supabaseUrl}/functions/v1/coach") {
            header("Authorization", "Bearer $token")
            header("apikey", Config.supabaseAnonKey)
            header("Content-Type", "application/json")
            setBody(buildJsonObject { put("question", question) }.toString())
        }
        val body = json.decodeFromString<JsonObject>(response.bodyAsText())
        // The function returns a readable reason for every refusal, so show it
        // rather than a generic failure.
        return body["answer"]?.jsonPrimitive?.content
            ?: body["error"]?.jsonPrimitive?.content
            ?: "No answer came back."
    }
}

File: App/CommissionersCartel/Features/Matchups/CoachViewModel.swift
import Foundation
import Observation

@Observable
@MainActor
final class CoachViewModel {
    struct Turn: Identifiable {
        let id = UUID()
        let text: String
        let isCoach: Bool
    }

    private(set) var turns: [Turn] = []
    private(set) var isBusy = false
    private var hasLoadedHistory = false

    /// Brings back what was said before. The conversation lives on the server,
    /// so it survives closing the app and follows a member between devices.
    func loadHistory(using environment: AppEnvironment) async {
        guard !hasLoadedHistory, environment.auth != nil else { return }
        hasLoadedHistory = true
        guard let history = try? await environment.content.coachHistory(limit: 100),
              !history.isEmpty else { return }
        // Anything already typed this session stays at the end, where it was.
        turns = history.map { Turn(text: $0.text, isCoach: $0.isCoach) } + turns
    }

    func ask(_ question: String, using environment: AppEnvironment) async {
        let trimmed = question.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !isBusy else { return }

        turns.append(Turn(text: trimmed, isCoach: false))
        isBusy = true
        defer { isBusy = false }

        turns.append(Turn(text: await answer(to: trimmed, using: environment), isCoach: true))
    }

    private func answer(to question: String, using environment: AppEnvironment) async -> String {
        guard
            let url = environment.configuration.supabaseURL?
                .appending(path: "/functions/v1/coach"),
            // The member's own token, not the anon key: the function answers
            // about whoever is asking, so it has to know who that is.
            let token = await environment.auth?.accessToken()
        else {
            return "Sign in first, under Settings."
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue(environment.configuration.supabaseAnonKey, forHTTPHeaderField: "apikey")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["question": question])

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let body = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            // The function gives a readable reason for every refusal, so show
            // it rather than a generic failure.
            return (body?["answer"] as? String)
                ?? (body?["error"] as? String)
                ?? "No answer came back."
        } catch {
            return "Coach Landry isn't answering just now."
        }
    }
}
